File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P17.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.AbstractMap.SimpleEntry;
import java.util.List;
import java.util.Map;
import java.util.stream.IntStream;

import static java.util.stream.Collectors.*;

/**
 * <b>(*) Split a list into two parts; the length of the first part is given</b>
 */
public class P17 {

    public static <T> Map<Boolean, List<T>> split(List<T> list, int n) {
        return IntStream
                .range(0, list.size())
                .mapToObj(i -> new SimpleEntry<>(i, list.get(i)))
                .collect(partitioningBy(entry -> entry.getKey() < n, mapping(SimpleEntry::getValue, toList())));
    }
}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P17.scala
package com.shekhargulati.ninetynine_problems._01_lists

/**
  * (*) Split a list into two parts; the length of the first part is given.
  * {{{
  *     split(List("a", "b", "c", "d", "e", "f", "g", "h", "i", "k"), 3)
  * }}}
  */
object P17 {

  def split[T](list: List[T], n: Int): (List[T], List[T]) = list.splitAt(n)

  def split_take_drop[T](list: List[T], n: Int): (List[T], List[T]) = (list.take(n), list.drop(n))

}
